src/object/srv_obj.c

```diff
@@ -612,6 +612,126 @@ ds_obj_rw_echo_handler(crt_rpc_t *rpc)
+		 int sgl_nr)
+{
+	struct ds_bulk_async_args arg = { 0 };
+	crt_bulk_opid_t		bulk_opid;
+	crt_bulk_op_t		bulk_op = CRT_BULK_GET;
+	crt_bulk_perm_t		bulk_perm = CRT_BULK_RW;
+	int			i, rc, *status, ret;
+
+	rc = ABT_eventual_create(sizeof(*status), &arg.eventual);
+	if (rc != 0)
+		return dss_abterr2der(rc);
+
+	for (i = 0; i < sgl_nr; i++) {
+		d_sg_list_t		*sgl, tmp_sgl;
+		struct crt_bulk_desc	 bulk_desc;
+		crt_bulk_t		 local_bulk_hdl;
+		struct bio_sglist	*bsgl;
+		daos_size_t		 offset = 0;
+		unsigned int		 idx = 0;
+		unsigned int		 sl_idx = 0;
+
+		if (remote_bulks[i] == NULL)
+			continue;
+
+		D_ASSERT(!daos_handle_is_inval(ioh));
+		bsgl = vos_iod_sgl_at(ioh, i);
+		D_ASSERT(bsgl != NULL);
+
+		sgl = &tmp_sgl;
+		rc = bio_sgl_convert(bsgl, sgl);
+		if (rc)
+			break;
+
+		if (daos_io_bypass & IOBP_SRV_BULK) {
+			/* this mode will bypass network bulk transfer and
+			 * only copy data from/to dummy buffer. This is for
+			 * performance evaluation on low bandwidth network.
+			 */
+			bulk_bypass(sgl, bulk_op);
+			goto next;
+		}
+
+		while (idx < sgl->sg_nr_out) {
+			d_sg_list_t	sgl_sent;
+			daos_size_t	length = 0;
+			unsigned int	start = idx;
+
+			sgl_sent.sg_iovs = &sgl->sg_iovs[start];
+			if (skip_list[i] == NULL) {
+				/* Find the end of the non-empty record */
+				while (sgl->sg_iovs[idx].iov_buf != NULL &&
+					idx < sgl->sg_nr_out)
+					length += sgl->sg_iovs[idx++].iov_len;
+			} else {
+				while (skip_list[i][sl_idx] < 0)
+					offset +=
+					(uint64_t)-skip_list[i][sl_idx++];
+				length += sgl->sg_iovs[idx++].iov_len;
+				D_ASSERT(skip_list[i][sl_idx] == length);
+				sl_idx++;
+			}
+			sgl_sent.sg_nr = idx - start;
+			sgl_sent.sg_nr_out = idx - start;
+
+			rc = crt_bulk_create(rpc->cr_ctx, &sgl_sent,
+					     bulk_perm, &local_bulk_hdl);
+			if (rc != 0) {
+				D_ERROR("crt_bulk_create %d error (%d).\n",
+					i, rc);
+				break;
+			}
+
+			crt_req_addref(rpc);
+			bulk_desc.bd_rpc	= rpc;
+			bulk_desc.bd_bulk_op	= bulk_op;
+			bulk_desc.bd_remote_hdl	= remote_bulks[i];
+			bulk_desc.bd_local_hdl	= local_bulk_hdl;
+			bulk_desc.bd_len	= length;
+			bulk_desc.bd_remote_off	= offset;
+			bulk_desc.bd_local_off	= 0;
+
+			arg.bulks_inflight++;
+			if (bulk_bind)
+				rc = crt_bulk_bind_transfer(&bulk_desc,
+					bulk_complete_cb, &arg, &bulk_opid);
+			else
+				rc = crt_bulk_transfer(&bulk_desc,
+					bulk_complete_cb, &arg, &bulk_opid);
+			if (rc < 0) {
+				D_ERROR("crt_bulk_transfer %d error (%d).\n",
+					i, rc);
+				arg.bulks_inflight--;
+				crt_bulk_free(local_bulk_hdl);
+				crt_req_decref(rpc);
+				break;
+			}
+			offset += length;
+		}
+next:
+		daos_sgl_fini(sgl, false);
+		if (rc)
+			break;
+		D_FREE(skip_list[i]);
+	}
+
+	if (arg.bulks_inflight == 0)
+		ABT_eventual_set(arg.eventual, &rc, sizeof(rc));
+
+	ret = ABT_eventual_wait(arg.eventual, (void **)&status);
+	if (rc == 0)
+		rc = ret ? dss_abterr2der(ret) : *status;
+
+	ABT_eventual_free(&arg.eventual);
+	return rc;
```

Comment:
this seems quite similar as ds_bulk_transfer() ?

Reply:
Yes, but it's different in that is only pulls specified sub-ranges. I looked at doing it within ds_bulk_transfer, but the ds_bulk_transfer function is already complicated by supporting both pulls (updates) and pushes (fetch).

Reply:
EC probably also need to do both put and get later, so maybe better to combine this one with ds_bulk_transfer (can do it in following patch).

Reply:
agree, after we make more progress on this, we should revisit this function and see if we can merge it with replication bulk transfer